### PR TITLE
Sync view configuration API updates from Gutenberg

### PR DESCRIPTION
**Pull request by oandregal**
Trac ticket: https://core.trac.wordpress.org/ticket/65577
Backports: https://github.com/WordPress/gutenberg/pull/80319
Follow-up to: https://github.com/WordPress/wordpress-develop/pull/12391

## What

Ports the latest iteration of the view configuration API from the Gutenberg update/view-config-api-versioning branch:

## How

- Rework the `WP_View_Config_Data` write API around `merge()`, `replace()`, `set()`, and `remove()`, all operating on patches of top-level keys and taking the schema version the change was authored against.
- Make `get_data()` private and move filter application into a new apply_filters() method so callbacks cannot read the materialized configuration and become coupled to its shape.
- Update the default post type configuration callbacks to use the new single-patch `set()` signature.
- Adapt and expand the unit tests accordingly.

## Use of AI Tools

AI assistance: Yes
Tool(s): Claude Code, Zed agent
Model(s): Opus, Fabble
Used for: Initial code skeleton and test suggestions; final implementation and tests were reviewed and edited by me.

### CODE REVIEW

**Review by jorgefilipecosta — APPROVED** [review]
There are some hedge cases that I would want to discuss further but nothing catastrophic or a blocker. Manually testing things seem to work well (during my smoke tests I did not found issues), there is extensive test coverage, so I think this is ready to ship.

File: src/wp-includes/class-wp-view-config-data.php [hunk 1]
```diff
@@ -71,88 +94,180 @@ class WP_View_Config_Data {
...
+		 *   but swaps each named value in wholesale, dropping whatever the key held
+		 *   before — for a callback that owns those keys outright.
+		 * - `remove( $spec, $version )` deletes named properties. The spec mirrors
+		 *   the configuration shape: a list of names deletes entries at that level,
+		 *   and a nested map recurses to prune from within a named value, down to
+		 *   individual list members.
+		 *
+		 * A change that declares an unsupported schema version is rejected and does
+		 * not alter anything. Callbacks mutate the container in place, so there is no
+		 * need to return it; any returned value is ignored. Callbacks must not replace
```
Comment (jorgefilipecosta) [inline src/wp-includes/class-wp-view-config-data.php:163]: This comment is wrong we need to return the data. I created a small doc fix PR at https://github.com/WordPress/wordpress-develop/pull/12641.

**Review by jorgefilipecosta — COMMENTED** [review]
I found some hedge cases while reviewing rather that going back and forth I proposed a fix for the cases I found at https://github.com/WordPress/gutenberg/pull/80571, https://github.com/WordPress/wordpress-develop/pull/12644. For each case I included a unit test, and left a comment with the output that was happening before the fix e.g: on this PR.
